samples: tfm: PSA firmware update sample

theotherjimmy opened the pull request:
Supersedes https://github.com/zephyrproject-rtos/zephyr/pull/34401

This commit adds a sample application demonstrating how to use
the PSA Firmware Update API from TF-M. It also enables the
FIRMWARE_UPDATE partition to be included at build time.

microbuilder committed:
samples: tfm: PSA firmware update sample

This commit adds a sample application demonstrating how to use
the PSA Firmware Update API from TF-M. It also enables the
`FIRMWARE_UPDATE` partition to be included at build time.

Signed-off-by: Kevin Townsend <kevin.townsend@linaro.org>
Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>

theotherjimmy committed:
samples: psa-firmware: Push firmware over psa_fwu_write

Previously, the example was a scheleton. This patch
pushes firmware images with psa_fwu_write, and completes
the firmware update example.

Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>

microbuilder commented on samples/tfm_integration/psa_firmware/CMakeLists.txt:
```diff
@@ -0,0 +1,102 @@
+cmake_minimum_required(VERSION 3.13.1)
+
+# Override the binary used by qemu
+set(QEMU_KERNEL_OPTION "-device;loader,file=${CMAKE_BINARY_DIR}/tfm_merged.hex")
```
This should be covered by the board cmake file(s) already:

- mps2_an521_ns: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mps2_an521/board.cmake#L15
- mps3_an547_ns: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mps3_an547/board.cmake#L24

theotherjimmy replied:
Done. Pushing soon.

microbuilder commented on samples/tfm_integration/psa_firmware/src/shell.c:
```diff
@@ -0,0 +1,40 @@
+static int
+psa_shell_cmd_version(const struct shell *shell, size_t argc, char **argv)
+{
+	shell_print(shell, "%s", "0.0.0");
```
Please update this to display the current firmware version(s).

theotherjimmy replied:
Done

microbuilder replied:
Can we also display the S and MCUBoot version numbers? I don't know if MCUBoot is available, but at least S is useful to know.

microbuilder commented:
@theotherjimmy Can you update this with a firmware image that increments the version number, to test rollback protection?

For the NS image, I believe this can be set via the config flag below, but it may be different when TF-M is enabled. (Please at a comment to the sample README with the right config flag to set the S and NS version numbers):

```
-DCONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS=\"--version=1.0.0\"
```

Also, perhaps we should update the pre-compiled firmware filename to some variation of this, to encompass other boards in the future (assuming v 1.0.1 here):

```
/boards/zephyr_mps3_an547_ns_1_0_1.hex
```

microbuilder commented:
This should also demonstrate how to handle a failed update, such as trying to apply an unsigned or incorrectly signed image, or a firmware image with an earlier version number if rollback protection is enabled

Assuming the default NS firmware version is set to 1.0.0, we'd try to apply these in order:

- zephyr_mps3_an547_ns_1_0_1_unsigned.hex (should be rejected)
- zephyr_mps3_an547_ns_0_1_0.hex (signed, but older version, should be rejected)
- zephyr_mps3_an547_ns_1_0_1.hex (signed and valid)

theotherjimmy commented:
@microbuilder I have added the last two tests, called failure-rollback and update-success respectively. I did not add the first one, as the example depends on the presence of a header (trailer?), so I could not find a way to have an unsigned image.

theotherjimmy commented:
Ah, compliance does not like intelhex files. It dislikes the line endings.

microbuilder commented:
@joerchan Mind doing a review?

theotherjimmy committed:
samples: psa-firmware: Document purpose of split-header.py

This should provide context to this small python script

Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>

theotherjimmy committed:
samples: psa-firmware: Display active Secure firmware version

This displays the secure firmware version before the nonsecure firmware
version at the beginning of boot.

Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>

microbuilder reviewed (changes requested):
Can we update this sample to set a version for the Secure image during build as well, to demonstrate how to set version numbers on both images? It's currently defaulting to 0.0.0-0 on the secure side:

```
Active S image version: 0.0.0-0
Active NS image version: 0.0.1-0
```

joerchan commented:
> Mind doing a review?

Yes, will have a look.

theotherjimmy committed:
modules-tfm: Configure image versions with KConfig

Previously, you were required to set the image versions through the
CMake variables TFM_IMAGE_VERSION_{S,NS}. For better integration with
the rest of the zephyr build system, these are now KConfig variables
with the same name.

Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>

theotherjimmy commented:
@microbuilder The last commit should do it.

theotherjimmy commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -343,12 +345,12 @@ if (CONFIG_BUILD_WITH_TFM)
-  if (NOT DEFINED TFM_IMAGE_VERSION_S)
-    set(TFM_IMAGE_VERSION_S 0.0.0+0)
+  if (NOT DEFINED CONFIG_TFM_IMAGE_VERSION_S)
+    set(CONFIG_TFM_IMAGE_VERSION_S 0.0.0+0)
```
I'll remove it

theotherjimmy commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -176,6 +176,22 @@ config TFM_BL2
 
 if TFM_BL2
 
+config TFM_IMAGE_VERSION_S
```
Okay. I'll move them above.

theotherjimmy commented on samples/tfm_integration/psa_firmware/CMakeLists.txt:
```diff
@@ -0,0 +1,105 @@
+)
+
+# Node (3) in the above graphic
+add_custom_command(
+  OUTPUT ${UPDATE_HEADER_BIN}
+  OUTPUT ${UPDATE_BIN}
+  DEPENDS ${UPDATE_SIGNED_HEX}
+  COMMAND ${PYTHON_EXECUTABLE}
+    ${CMAKE_CURRENT_LIST_DIR}/split-header.py
+    ${UPDATE_SIGNED_HEX}
+    ${UPDATE_BIN}
+    ${UPDATE_HEADER_BIN}
+)
+
+# Node (4) in the above graphic
+add_custom_command(
+  OUTPUT ${UPDATE_HEADER_OBJ}
+  DEPENDS ${UPDATE_HEADER_BIN}
+  COMMAND ${CMAKE_OBJCOPY} -I binary -O elf32-littlearm -B arm
+    ${UPDATE_HEADER_BIN}
+    ${UPDATE_HEADER_OBJ}
+)
```
I don't think that's needed. It seems to work as is, including only generating the needed files.

theotherjimmy replied:
Seems that there's no problem when they're in the same directory? and only for the Makefile backend.

theotherjimmy replied:
I used that particular use case when debugging an issue with `split-headers.py`, and did not see any issues.

theotherjimmy commented on modules/trusted-firmware-m/Kconfig.tfm:
```diff
@@ -211,6 +227,16 @@ config TFM_MCUBOOT_PATH_DOWNLOAD
 
 endchoice
 
+config TFM_PARTITION_FIRMWARE_UPDATE
```
Moved into the `.partitions` file

joerchan commented on samples/tfm_integration/psa_firmware/README.rst:
```diff
@@ -0,0 +1,187 @@
+      cd <ZEPHYR_ROOT>/samples/tfm_integration/psa_firmware/
+      rm -rf build
+      mkdir build && cd build
+      cmake -GNinja -DBOARD=mps2_an521_ns ..
```
Use `mps3_an547_ns` here as well

theotherjimmy replied:
Ah, oops.

joerchan commented on samples/tfm_integration/psa_firmware/README.rst:
```diff
@@ -0,0 +1,187 @@
+===============
+
+1. Build Zephyr with a non-secure configuration
+   (``-DBOARD=mps2_an521_ns``).
```
Should be mps3_an547_ns, since title is MPS3 AN547

joerchan commented on samples/tfm_integration/psa_firmware/README.rst:
```diff
@@ -0,0 +1,187 @@
+
+   .. code-block:: bash
+
+      cd <ZEPHYR_ROOT>
```
There are shorthands for this:
```
.. zephyr-app-commands::
   :zephyr-app: samples/tfm_integration/psa_firmware
   :host-os: unix
   :board: mps3_an547_ns
   :goals: run
   :compact:
```

joerchan commented on samples/tfm_integration/psa_firmware/README.rst:
```diff
@@ -0,0 +1,187 @@
+      cd <ZEPHYR_ROOT>
+      west build -p -b mps3_an547_ns samples/tfm_integration/psa_firmware -t run
+
+   Using ``cmake`` and ``ninja``
```
I think we have stopped documenting using cmake ninja/make, and only document using west commands.

Maybe we shouldn't have these?

theotherjimmy replied:
Happy to remove them.

theotherjimmy committed:
samples: psa-firmware: Correct readme instructions

Prior instructions were incomplete, included extra, unneeded ways to
build the app.

This should simplify the instructions.

Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>

microbuilder commented:
@joerchan @sigvartmh Mind having another look? Thanks.

d3zd3z commented on samples/tfm_integration/psa_firmware/boards/hello-an547.hex:
```diff
@@ -0,0 +1,3614 @@
+:020000040106F3
```
How difficult would it be to Build and sign this during CI rather than checking in the hex file?

microbuilder replied:
Probably a case for `extra_post_build_commands` in the project cmake file? (Ex.: https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/trusted-firmware-m/CMakeLists.txt#L458) ... but given that this file exists purely for CI, personally I'm OK with having the pre-built image unless you feel strongly we should automate this.

theotherjimmy replied:
It has to be built _before_ your build. I'll punt on this for now. If you would like, opening an issue to improve this would be fine with me.

d3zd3z reviewed (dismissed):
Just a minor question.

Also, if you do redo the changes, "scheleton" -> "skeleton" in one of the commit messages.

joerchan commented:
@theotherjimmy Did you consider the question I had about MCUBOOT_DATA_SHARING=ON?

I'm wondering if this should be a different Kconfig. And TFM_BL2 should select it instead.
If we don't have firmware update partition then we might not want this enabled.

microbuilder commented:
> @theotherjimmy Did you consider the question I had about MCUBOOT_DATA_SHARING=ON?
> 
> I'm wondering if this should be a different Kconfig. And TFM_BL2 should select it instead. If we don't have firmware update partition then we might not want this enabled.

I think this is a sensible approach, yes. Thanks for pointing it out again.

theotherjimmy committed:
tfm: Kconfig: Configure MCUBOOT_DATA_SHARING explicitly

Previously, MCUBOOT_DATA_SHARING was enabled when BL2 was built
and when the firmware update partition was present. This is not
the only situation that you might be interested in this data
sharing. The data sharing now has it's own Kconfig.

Further, use of the firmware update partition requires data
sharing, so it selects it.

Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>

theotherjimmy commented:
@microbuilder @joerchan Included an MCUBOOT_DATA_SHARING config in a new commit. Tests passed locally on my machine, so CI will probably pass too.

joerchan commented:
> @microbuilder @joerchan Included an MCUBOOT_DATA_SHARING config in a new commit. Tests passed locally on my machine, so CI will probably pass too.

Does the FIRMWARE_UPDATE partition require this config? If that is the case then a select is better:
```
config TFM_PARTITION_FIRMWARE_UPDATE
	bool "Include the secure parition 'Firmware Update'"
        select MCUBOOT_DATA_SHARING
	default n
	help
```
I see you put it in the prj.conf instead. Which is fine if this is not a required configuration by the FIRMWARE_UPDATE partition.

microbuilder commented:
> ```
> config TFM_PARTITION_FIRMWARE_UPDATE
> 	bool "Include the secure parition 'Firmware Update'"
>         select MCUBOOT_DATA_SHARING
> 	default n
> 	help
> ```

I guess you meant `select TFM_MCUBOOT_DATA_SHARING`, but good point.

theotherjimmy commented:
Failure looks unrelated to me.

theotherjimmy commented:
Depends on https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/69.

theotherjimmy commented:
I'll rebase this on top of main after https://github.com/zephyrproject-rtos/zephyr/pull/44419 is merged.
